[docs] Add orcl to docusaurus (#6175)
magma/magma: docs/readmes/orc8r/deploy_orcl.md

---
id: deploy_orcl
title: Experimental Orc8r Deployer
hide_title: true
---

# Orcl

Orcl is a Orchestrator CLI. It is used for managing Orc8r deployment. It provides following subcommands.

- Configure
- Certs
- Install
- Upgrade
- Verify
- Cleanup
- Debug(perhaps in the future)

![Orcl Big Picture](assets/orc8r/orcl.png)
Orcl is packaged within orc8r_deployer. Orc8r deployer is a docker image which contains all the necessary prerequisites to deploy orc8r. The only requirements for running orc8r_deployer is that the the host machine must have [docker engine installed](https://docs.docker.com/get-docker/).

## Usage

```
./run_deployer runs the orc8r deployer container

> orc8r deployer contains scripts which enable user to configure, run prechecks

Reviewer: Missing a comma at the end of this line
```suggestion
orc8r deployer contains scripts which enable user to configure, run prechecks,
```

install, upgrade, verify and cleanup an orc8r deployment
Usage: run_deployer [-deploy-dir|-root-dir|-build|-h]
options:
-h           Print this Help
-deploy-dir  deployment dir containing configs and secrets (mandatory)
-root-dir    magma root directory
-build       build the deployer container
example: ./run_deployer -deploy-dir /tmp/orc8r_14_deployment

> ```

Reviewer: Some more whitespace can be helpful here. It's a little difficult to parse at a glance

Reviewer: Not a comment about the docs, specifically, but options that are longer than a single letter typically have two hyphens.

```
--help
-h
--deploy-dir
-d
--root-dir
-r
--build
-b
```

Author: I think both styles are used. For e.g. in terraform there is only single dashed option. That being said, The orcl cli internally uses click which uses double hyphen for multiple letters. I will change this and keep it consistent.

> Deployment directory will be used to maintain the configs, secrets and also contain the deployment related files such as main.tf etc. It is important to use the same deployment directory during the upgrade so that the configuration variables can be reused.

Reviewer: Should `orcl` be used with specifying a `/tmp` directory for a more permanent deployment? The earlier section is a little bit misleading with that.

Author: It is an example. I can set some other location if you thought this was misleading.

> In the following section, we will discuss each of the orcl commands in detail.

Reviewer: ```suggestion
In the following sections, we will discuss each of the orcl commands in detail.
```

> ## Configure

Reviewer: I think these headers should explicitly state these are subcommands - just a little bit easier when skimming
```suggestion
## Configure Command
```

> 

Reviewer: I liked how the other sections begin with a clear statement of what the command is for. Copying and pasting from the paragraph in the next line:
```suggestion
Configure command enables user to easily configure the mandatory configs necessary for the deployment.
```

Author: This is the entire description  of the command.
"Every orc8r deployment relies on several configuration attributes. For example, cluster_name attribute is used to identify the orc8r  kubernetes cluster, orc8r_tag provides the image tag to be used during the deployment. Configure command enables user to easily configure the mandatory configs necessary for the deployment. It additionally provides the ability to also configure optional attributes through the ***set*** subcommand. Configure command also has subcommands like ***info*** to show all the possible configuration attributes and ***show*** to display the current configuration. Finally configure also contains a subcommand ***check*** which provides the ability to check if all mandatory configs needed by the deployment have been successfully configured."
The configure command has subcommands set, show, info etc. So I had explained everything in there.

> Every orc8r deployment relies on several configuration attributes. For e.g cluster_name attribute is used to identify the orc8r  kubernetes cluster, orc8r_tag provides the image tag to be used during the deployment. Configure command enables user to easily configure the mandatory configs necessary for the deployment. It additionally provides the ability to also configure optional attributes through the ***set*** subcommand. Configure command also has subcommands like ***info*** to show all the possible configuration attributes and ***show*** to display the current configuration. Finally configure also contains a subcommand ***check*** which provides the ability to check if all mandatory configs needed by the deployment have been successfully configured.

Reviewer: ```suggestion
Every orc8r deployment relies on several configuration attributes. For example cluster_name attribute is used to identify the orc8r kubernetes cluster, orc8r_tag provides the image tag to be used during the deployment. Configure command enables user to easily configure the mandatory configs necessary for the deployment. It additionally provides the ability to also configure optional attributes through the ***set*** subcommand. Configure command also has subcommands like ***info*** to show all the possible configuration attributes and ***show*** to display the current configuration. Finally configure also contains a subcommand ***check*** which provides the ability to check if all mandatory configs needed by the deployment have been successfully configured.
```

  Configure orc8r deployment variables

Options:
  -c, --component [infra|platform|service]
  --help                          Show this message and exit.

Commands:
  check  Check if all mandatory configs are present
  info   Display all possible config options in detail
  set    Set enables user to configure any configuration option.
  show   Display the current configuration
```

Note: Currently we don’t have the capability to sanitize the inputs. This might be a feature to be added later

Configure command relies mainly on a meta file `vars.yml`. This file contains all deployment variables with their name, type, description, default value, whether it is a required or an optional attribute and finally the apps which rely on this configuration attribute.
The configuration provided by the user is used to

- set the aws configuration(specifically access_key, secret_key, region),
- build terraform.tfvars.json(which is automatically loaded by terraform and provided to the root module)
- build main.tf and vars.tf(to ensure that we set the vars in the root module for it to be correspondingly consumed by the orc8r and orc8r-app modules)

> For e.g.

Reviewer: Preferentially use the English word over the latin abbreviation except under narrow circumstances

```suggestion
For example
```

```
module "orc8r" {
  source = "/root/magma/orc8r/cloud/deploy/terraform/orc8r-aws"
  cluster_name=var.cluster_name <-- generated from configured keys
  cluster_version=var.cluster_version
  orc8r_domain_name=var.orc8r_domain_name
  region=var.region
```

## Certs

> Certs is a orcl command for managing certificates. Currently it has only one subcommand ***add*** which can be used to add application certs and self signed root certificates(optional).

Reviewer: For quick referencing, subsections can be added for each command to list and explain the subcommands.

Author: I have added that already below. 
```

Commands:
  add  Add creates application and self signed(optional) certs
```

**Note: it is important to use the same deployment directory which was used during install. This is mainly to ensure that the terraform state is available and also reuse the configuration variables**

```
# orcl upgrade --help
Usage: orcl upgrade [OPTIONS] COMMAND [ARGS]...

  Upgrade existing orc8r deployment

Options:
  --help  Show this message and exit.

Commands:
  precheck  Precheck runs various checks to ensure successful upgrade
```

Upgrade command runs the following upgrade command

```
terraform apply
```

Following prechecks are performed prior to the upgrade

- Check if terraform state exists
- Check if all mandatory configs are present
- Check if we have atleast 3 availability zones in our deployment
- Check if deployed EKS cluster version is greater than what’s specified in the configuration
- In case of elastic search deployment, check if AWSServiceRoleForAmazonElasticsearchService IAM role already exists
- Check if deployed RDS instance version is greater than what’s specified in the configuration
- Check if image tag specified is present in image repository
- Check if helm chart specified is present in the helm repository

## Verify

> Verify command provides the ability to post deployment checks on orc8r.

Reviewer: ```suggestion
Verify command provides the ability to run post-deployment checks on orc8r.
```

```
Usage: orcl verify [OPTIONS] COMMAND [ARGS]...

>   Run post deployment checks on orc8r

Reviewer: ```suggestion
  Run post-deployment checks on orc8r
```
